tests/functionaltests/Tests/UIKitTests/UIViewControllerTests.mm

```diff
@@ -0,0 +1,310 @@
+            EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidAppear:)));
+        }
+    }
+}
+
+- (void)viewWillDisappear:(BOOL)animated {
+    [super viewWillDisappear:animated];
+
+    EXPECT_TRUE(sel_isEqual([self lastEvent], @selector(viewDidAppear:)));
+    _lastEvent = _cmd;
+
+    if ([self presentedViewController]) {
+        // About to be covered by presented modal
+
+        EXPECT_FALSE([self isBeingDismissed]);
+
+        ModalEventsTestViewController* presented = (ModalEventsTestViewController*)[self presentedViewController];
+
+        EXPECT_TRUE(_hidesParent(presented.modalPresentationStyle));
+
+        // We should receive viewWillDisappear: before presented receives viewWillAppear:
+        // Ensure presented is displaying for the first time (or was previously dismissed)
+        EXPECT_TRUE(sel_isEqual([presented lastEvent], @selector(viewDidLoad)) ||
+                    sel_isEqual([presented lastEvent], @selector(viewDidDisappear:)));
+    } else {
+        // About to be dismissed
+
+        EXPECT_TRUE([self isBeingDismissed]);
+
+        EXPECT_NE([self presentingViewController], nil);
+
+        if ([[self presentingViewController] isKindOfClass:[ModalEventsTestViewController class]]) {
+            ModalEventsTestViewController* presenting = (ModalEventsTestViewController*)[self presentingViewController];
+
+            if (_hidesParent(self.modalPresentationStyle)) {
+                // We should receive viewWillDisappear: before presenting receives viewWillAppear:
+                EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidDisappear:)));
+            } else {
+                EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidAppear:)));
+            }
+        }
+    }
+}
+
+- (void)viewDidDisappear:(BOOL)animated {
+    [super viewDidDisappear:animated];
+
+    EXPECT_TRUE(sel_isEqual(_lastEvent, @selector(viewWillDisappear:)));
+    _lastEvent = _cmd;
+
+    if ([self presentedViewController]) {
+        // Modal is covering us.
+
+        EXPECT_FALSE([self isBeingDismissed]);
+
+        EXPECT_NE([self presentingViewController], nil);
+
+        ModalEventsTestViewController* presented = (ModalEventsTestViewController*)[self presentedViewController];
+
+        EXPECT_TRUE(_hidesParent(presented.modalPresentationStyle));
+
+        // presented should receive viewDidAppear: first.
+        EXPECT_TRUE(sel_isEqual([presented lastEvent], @selector(viewDidAppear:)));
+    } else {
+        // We're a dismissed modal.
+        EXPECT_EQ([self presentingViewController], nil);
+
+        EXPECT_TRUE([self isBeingDismissed]);
+    }
+}
+
+@end
+
+class UIViewControllerModalTest : public ::testing::TestWithParam<::testing::tuple<UIModalPresentationStyle, WOCOperationMode, BOOL>> {
+protected:
+    virtual void SetUp() {
+        FunctionalTestSetupUIApplication();
+    }
+
+    virtual void TearDown() {
+        FunctionalTestCleanupUIApplication();
+    }
```

Comment:
Ran into some issues changing that. If I follow the BEGIN_TEST_CLASS/TEST_CLASS_SETUP pattern while still inheriting from the TestWithParam template class I'll end up with

    allocating an object of abstract class type 'UIViewControllerModalTest'	FunctionalTests (Tests\FunctionalTests\FunctionalTests)	C:\Users\emetcalfe\.nuget\packages\Taef.Redist.Wlk\1.0.170206001-nativeTargets\build\Include\WexTestClass.h	440	

Also looking at the INSTANTIATE_TEST_CASE_P macro https://github.com/Microsoft/WinObjC/blob/650e74942ccf99907aac59f250df9151ec34a70d/tests/Tests.Shared/test-api.h#L500 it expects a class with methods named SetUp and TearDown, and I don't think these are created by the TEST_CLASS_SETUP macro in WexTestClass.h

One option would be to just add an explicit TEST_METHOD for each test case and avoid the parameterization via INSTANTIATE_TEST_CASE_P. I still have a version of the patch that does this so it's easy to make this change. It would also provide more clear test names e.g. TestPresentDismissPopoverTabletNonAnimated instead of just a numbered test name.